std/algorithm/comparison.d

```diff
@@ -1567,3 +1567,156 @@ unittest
 }
+
+/** Returns: First Argument (element of $(D a)) whose implicit conversion to
+    bool is true.
+
+    Similar to behaviour of or operator in dynamic languages such as Lisp's (or
+    a...) and Python's a or ....
+
+    TODO Is inout Conversion!T the correct return value?
+
+    NOTE: Lazy parameters are currently marked as throw which makes it
+    impossible to mark either() as nothrow. Issue at
+    https://issues.dlang.org/show_bug.cgi?id=12647
+*/
+CommonType!Ts either(Ts...)(lazy Ts a) if (a.length >= 1)
+{
+    auto a0 = a[0]();           // evaluate only once
+    static if (Ts.length == 1)
+    {
+        return a0;
+    }
+    else
+    {
+        return a0 ? a0 : either(a[1 .. $]); // recurse
+    }
+}
+
+/** This overload enables, when possible, lvalue return.
+    TODO should we limit this template to a.length >= 2
+*/
+auto ref either(Ts...)(ref Ts a) if (a.length >= 1 && allSame!Ts)
+{
+    static if (Ts.length == 1)
+    {
+        return a[0];
+    }
+    else
+    {
+        return a[0] ? a[0] : either(a[1 .. $]); // recurse
+    }
+}
+
+@safe pure unittest
+{
+    immutable p = 1, q = 2;
+    auto pq = either(p, q);
+    assert(pq == 1);
+
+    assert(either(3) == 3);
+    assert(either(3, 4) == 3);
+    assert(either(0, 4) == 4);
+    assert(either(0, 0) == 0);
+    assert(either("", "a") == "");
+    string s = null;
+    assert(either(s, "a") == "a");
+    assert(either("a", "") == "a");
+    immutable a2 = [1, 2];
+    assert(either(a2) == a2);
+    assert(either([0, 1], [1, 2]) == [0, 1]);
+    assert(either([0, 1], [1]) == [0, 1]);
+    assert(either("a", "b") == "a");
+
+    int x = 1, y = 2;
+    either(x, y) = 3;
+    assert(x == 3);
+    assert(y == 2);
+}
+
+/** Returns: Last Argument if all arguments implicitly bool-convert to true
+    otherwise CommonType!T.init.
+
+    Similar to behaviour and operator in dynamic languages such as of Lisp's
+    (and a...) and Python's a and ....
+
+    TODO Is inout Conversion!T the correct return value?
+*/
+CommonType!T every(T...)(lazy T a) if (T.length >= 1)
+{
+    auto a0 = a[0]();           // evaluate only once
+    static if (T.length == 1)
+    {
+        return a0;
+    }
+    else
+    {
+        return a0 ? every(a[1 .. $]) : CommonType!T.init; // recurse
+    }
+}
+
+@safe pure unittest
+{
+    assert(every(3) == 3);
+    assert(every(3, 4) == 4);
+    assert(every(0, 4) == 0);
+    assert(every(0, 0) == 0);
+    assert(every([0, 1], [1, 2]) == [1, 2]);
+    assert(every([0, 1], [1]) == [1]);
+    assert(every("a", "b") == "b");
+    assert(every("", "b") == "b");
+    assert(every(cast(string)null, "b") == cast(string)null);
+}
+
+/** Evaluate all $(D parts) possibly digesting $(D whole).
+    If all values of $(D parts) implicitly convert to bool true return the
+    values as an array, otherwise restore whole and return null.
+*/
+bool[T.length] tryEvery(S, T...)(ref S whole,
+                                 lazy T parts) if (T.length >= 1)
```

Comment:
A better naming might be `skipOverAllOrNone` or perhaps just `skipOverAll` or perhaps just making skipOver `variadic`.